h2o-automl/src/main/java/ai/h2o/automl/AutoML.java

```diff
@@ -164,76 +166,142 @@ public AutoML() {
-  }
+  private void validateBuildSpec(AutoMLBuildSpec buildSpec) {
+    // Validate all buildSpec parameters here and provide reasonable defaults or parameter cleaning if necessary.
+    // When assigning default or modifying the original value, use `eventLog` to inform the user about the changes.
+    // Please no additional logic!
 
-  private void handleReproducibilityParameters(AutoMLBuildSpec buildSpec) {
-    eventLog().info(Stage.Workflow, "Build control seed: " + buildSpec.build_control.stopping_criteria.seed() +
-            (buildSpec.build_control.stopping_criteria.seed() == -1 ? " (random)" : ""));
+    // Can possibly use this validation entry point to provide parameters validation for Flow later.
+    // Can also reuse ModelBuilder.ValidationMessage for this.
+
+    validateInput(buildSpec.input_spec);
+    validateModelValidation(buildSpec);
+    validateModelBuilding(buildSpec.build_models);
+    validateEarlyStopping(buildSpec.build_control.stopping_criteria, buildSpec.input_spec);
+    validateReproducibility(buildSpec);
   }
 
-  private void handleEarlyStoppingParameters(AutoMLBuildSpec buildSpec) {
-    if (buildSpec.build_control.stopping_criteria.stopping_tolerance() == AUTO_STOPPING_TOLERANCE) {
-      buildSpec.build_control.stopping_criteria.set_default_stopping_tolerance_for_frame(_trainingFrame);
-      eventLog().info(Stage.Workflow, "Setting stopping tolerance adaptively based on the training frame: " +
-              buildSpec.build_control.stopping_criteria.stopping_tolerance());
-    } else {
-      eventLog().info(Stage.Workflow, "Stopping tolerance set by the user: " + buildSpec.build_control.stopping_criteria.stopping_tolerance());
-      double default_tolerance = AutoMLBuildSpec.AutoMLStoppingCriteria.default_stopping_tolerance_for_frame(_trainingFrame);
-      if (buildSpec.build_control.stopping_criteria.stopping_tolerance() < 0.7 * default_tolerance){
-        eventLog().warn(Stage.Workflow, "Stopping tolerance set by the user is < 70% of the recommended default of " + default_tolerance + ", so models may take a long time to converge or may not converge at all.");
+  private void validateInput(AutoMLInput input) {
+    if (null == DKV.getGet(input.training_frame))
+      throw new H2OIllegalArgumentException("No training data has been specified, either as a path or a key, or it is not available anymore.");
+
+    final Frame trainingFrame = DKV.getGet(input.training_frame);
+    final Frame validationFrame = DKV.getGet(input.validation_frame);
+    final Frame blendingFrame = DKV.getGet(input.blending_frame);
+    final Frame leaderboardFrame = DKV.getGet(input.leaderboard_frame);
+
+    Map<String, Frame> compatibleFrames = new LinkedHashMap<String, Frame>(){{
+      put("training", trainingFrame);
+      put("validation", validationFrame);
+      put("blending", blendingFrame);
+      put("leaderboard", leaderboardFrame);
+    }};
```

Comment:
@seb-h2o this double bracket syntax looks convenient but FYI I have heard that there are some concern about it:
```
Note that, we must try to avoid this initialization technique as it creates an anonymous extra class at every usage, and also holds hidden references to the enclosing object and might cause memory leak issues.
``` 
https://www.baeldung.com/java-initialize-hashmap

Reply:
the instance scope is only local to the validation method, so it's safe: no leak concerns.

Reply:
It's just the same concerns as for any anonymous class btw.